agentos/component.py

```diff
@@ -0,0 +1,252 @@
+        cls = getattr(module, class_name)
+        sys.path.pop()
+        return Component(
+            managed_cls=cls,
+            dependencies=dependencies,
+            requirements=requirements,
+            dunder_name=dunder_name,
+        )
+
+    @classmethod
+    def get_from_class(
+        cls,
+        managed_cls: typing.Type[T],
+        dependencies: typing.Dict[str, str] = None,
+        requirements: typing.List[str] = None,
+        dunder_name: str = "__component__",
+    ) -> "Component":
+        return Component(
+            managed_cls=managed_cls,
+            dependencies=dependencies,
+            requirements=requirements,
+            dunder_name=dunder_name,
+        )
+
+    @property
+    def name(self):
+        return self.managed_cls.__name__
+
+    def call(self, fn_name: str):
+        self.instantiate_class()
+        fn = getattr(self.instance, fn_name)
+        assert fn is not None, f"{self.instance} has no function {fn_name}"
+        params = self.params.get(fn_name, {})
+        print(f"Calling {self.name}.{fn_name}(**{params})")
+        return fn(**params)
+
+    def add_params(self, fn_name: str, params: typing.Dict):
+        print(f"{self.name}: adding {params} to {fn_name}()")
+        curr_params = self.params.get(fn_name, {})
+        curr_params.update(params)
+        self.params[fn_name] = curr_params
+
+    def add_dependency(self, component: "Component", alias: str = None):
+        if type(component) is not type(self):
+            raise Exception("add_dependency() must be passed a Component")
+        if alias is None:
+            alias = component.name
+        self.dependencies[alias] = component
+        self.dependency_names[alias] = component.name
+
+    def instantiate_class(self) -> None:
+        if self.instance is not None:
+            return
+        assert len(self.dependencies) == len(self.dependency_names)
+        save_init = self.managed_cls.__init__
+        self.managed_cls.__init__ = lambda self: None
+        self.instance = self.managed_cls()
+        for dep_alias, dep_component in self.dependencies.items():
+            dep_component.instantiate_class()
+            setattr(self.instance, dep_alias, dep_component.instance)
+        setattr(self.instance, self.dunder_name, self)
+        self.managed_cls.__init__ = save_init
+        self.call("__init__")
+
+    def get_instance(self) -> T:
+        self.instantiate_class()
+        return self.instance
+
+
+# A REPL-ready demo
+if __name__ == "__main__":
+
+    class SimpleAgent:
+        def __init__(self):
+            env_name = self.env.__class__.__name__
+            print(f"SimpleAgent: AgentOS added self.env: {env_name}")
+
+        def reset_env(self):
+            self.env.reset()
+
+    class SimpleEnvironment:
+        def reset(self):
+            print("SimpleEnvironment.reset() called")
+
+    # Generate Components from Classes
+    agent_component = Component.get_from_class(SimpleAgent)
```

Comment:
As we discussed today, IMO, the easiest to learn interface would be my_component = Component(), and then we could essentially have `__init__()` be an alias for `get_from_class()`